src/axom/slam/ProductSet.hpp

```diff
@@ -34,24 +36,81 @@ namespace slam
  *
  * \see   BivariateSet
  */
-template <typename SetType1 = slam::Set<>, typename SetType2 = slam::Set<>>
+template <typename SetType1 = slam::Set<>,
+          typename SetType2 = slam::Set<>,
+          typename InterfaceType = policies::VirtualInterface>
 class ProductSet final
-  : public BivariateSet<SetType1, SetType2>,
-    RangeSet<typename SetType1::PositionType, typename SetType1::ElementType>
+  : public policies::BivariateSetInterface<InterfaceType, SetType1, SetType2>
 {
+private:
+  using BaseType =
+    policies::BivariateSetInterface<InterfaceType, SetType1, SetType2>;
+
 public:
-  using RangeSetType =
-    RangeSet<typename SetType1::PositionType, typename SetType1::ElementType>;
+  using RangeSetType = typename BaseType::RangeSetType;
   using FirstSetType = SetType1;
   using SecondSetType = SetType2;
-  using BivariateSetType = BivariateSet<FirstSetType, SecondSetType>;
-  using PositionType = typename BivariateSetType::PositionType;
-  using ElementType = typename BivariateSetType::ElementType;
-  using OrderedSetType = typename BivariateSetType::OrderedSetType;
+  using PositionType = typename BaseType::PositionType;
+  using ElementType = typename BaseType::ElementType;
   using ProductSetType = ProductSet<SetType1, SetType2>;
 
+private:
+  template <typename Dummy, typename SetType>
+  struct RowSet
+  {
+    using Type = SetType;
+    RowSet(PositionType secondSetSize) : m_data(secondSetSize)
+    {
+      //fill in the row data now for getElements(i) function,
+      //since every row is the same, a call to getElements() returns the same set.
+      //
+      // HACK -- this should actually be returning a PositionSet since it always
+      //         goes from 0 to secondSetSize()
+      // This requires a change to the return type of BivariateSet::getElements()
+      std::iota(m_data.begin(), m_data.end(), 0);
+      m_set =
+        typename SetType::SetBuilder().size(secondSetSize).offset(0).data(&m_data);
+    }
+
+    Type get(PositionType) const { return m_set; }
+
+    std::vector<PositionType> m_data;
+    SetType m_set;
+  };
+
+  // This dummy parameter works around a C++ bug where partial specializations,
+  // but not explicit instantiations, are allowed in class scope.
+  // This is fixed in C++17: https://wg21.cmeerw.net/cwg/issue727
+  template <typename Dummy>
+  struct RowSet<Dummy, void>
+  {
+    using Type = PositionSet<PositionType, ElementType>;
+
+    RowSet(PositionType) { }
+
+    Type get(PositionType secondSetSize) const { return Type(secondSetSize); }
+  };
+
+public:
+  using ConcreteSet = ProductSet<SetType1, SetType2, policies::ConcreteInterface>;
+  using PolymorphicSet =
```

Comment:
Can you explain the difference between Polymorphic vs Virtual here? Is there a reason its not VirtualSet?

Reply:
Polymorphic sounds fancier :P

Point taken, I can change it to be consistent with Virtual[thing]Policy.